p2p/netaddress.go

```diff
@@ -207,22 +205,28 @@ func (na *NetAddress) DialTimeout(timeout time.Duration) (net.Conn, error) {
 
 // Routable returns true if the address is routable.
 func (na *NetAddress) Routable() bool {
+	if err := na.Valid(); err != nil {
+		return false
+	}
 	// TODO(oga) bitcoind doesn't include RFC3849 here, but should we?
-	return na.Valid() && !(na.RFC1918() || na.RFC3927() || na.RFC4862() ||
+	return !(na.RFC1918() || na.RFC3927() || na.RFC4862() ||
 		na.RFC4193() || na.RFC4843() || na.Local())
 }
 
 // For IPv4 these are either a 0 or all bits set address. For IPv6 a zero
 // address or one that matches the RFC3849 documentation address format.
-func (na *NetAddress) Valid() bool {
-	if string(na.ID) != "" {
-		data, err := hex.DecodeString(string(na.ID))
-		if err != nil || len(data) != IDByteLength {
-			return false
-		}
+func (na *NetAddress) Valid() error {
```

Comment:
the fact that NetAddress can be invalid bothers me deeply... I think we should refactor the code so NetAddress is always correct. For that we'll need to change NewNetAddressX funcs to return an error + make NewNetAddressIPPort private. Thoughts?

Reply:
I do agree that it should always be correct, but a validity check is always good. Should we open an issue and quickly discuss there, with others then move forward? or would you like to include those changes here?

Reply:
https://github.com/tendermint/tendermint/issues/3782